### PR TITLE
Remove and add latest versions of jquery, jquery-plugins, and backbone

### PR DESCRIPTION
**Pull request by kdjohnson**
I removed all previous versions of jquery, backbone, and the various jquery plugins. I also added in the latest if it was not present before.

### CODE REVIEW

**Comment by bpowell** [comment]
The UPSC said they wanted the old versions removed. Could @apetro @drewwills @bjagg @jimhelwig @asgrant chime in on this please.

**Comment by kdjohnson** [comment]
:+1:

**Comment by auxepaul** [comment]
uPortal and portlets won't work with this commit... becauce they don't use the last versions of jquery, jquey-migrate, jquery-ui, jquery-mobile,... (for example, uPortal will never use jquery-mobile 1.4.x...). I think you should keep at least **the last versions used by uPortal 4.3.x and uPortal 5.0** (jquery 1.11.x, jquery-ui 1.10.x, jquery-mobile 1.3.x... : https://github.com/Jasig/uPortal/blob/19c611fe3d27b53171e500e7c51fe4f24ca20bc5/uPortal-webapp/src/main/webapp/media/skins/respondr/common/common_skin.xml, https://github.com/Jasig/uPortal/blob/rel-4-3-patches/uportal-war/src/main/webapp/media/skins/muniversality/common/common_skin.xml) before breaking all community projects (or it is necessary to create two branches in resource-server projet).

**Comment by ChristianMurphy** [comment]
@auxepaul 

> uPortal and portlets won't work with this commit

There may be some additional upgrades needed across the uPortal ecosystem. ⬆️ 
However we need to remove dependencies on outdated and insecure libraries. ✅ 

---

> the last versions used by uPortal 4.3.x and uPortal 5.0

If this PR is considered a breaking change, then according to Semantic Versioning with would need to be in a major release (uPortal >=5).

> Major version X (X.y.z | X > 0) MUST be incremented if any backwards incompatible changes are introduced to the public API. It MAY include minor and patch level changes. Patch and minor version MUST be reset to 0 when major version is incremented.

:gb: http://semver.org/#spec-item-8
:fr: http://semver.org/lang/fr/#spec-item-8

Changes to uPortal 4.3 are limited by

> Patch version Z (x.y.Z | x > 0) MUST be incremented if only backwards compatible bug fixes are introduced. A bug fix is defined as an internal change that fixes incorrect behavior.

:gb: http://semver.org/#spec-item-6
:fr: http://semver.org/lang/fr/#spec-item-6

**Comment by auxepaul** [comment]
All used js libs are outdated in the "jasig projects/uPortal ecosystem" on Github ; it's a huge work that "no-one" is doing regularly (we try do that at UPMC when we put a new portal in production). So, "Good luck!"
I know and understand Semantic versioning but a best choice should be here : 
"First step" : Update all js dependencies (and others) on "jasig projects" (**All security updates must be immediate**)
"Second step" : Remove dependencies in resource-server (or create two temporary branches in order to facilitate the dependencies updates of uPortal ecosystem and the updates of old projects by the community).

Now, I think I'll close all jira tickets (that I writed and pushed) because our team works shouldn't work with this last commit.

**Comment by ChristianMurphy** [comment]
Hey @auxepaul :wave: 

### TL;DR

I expect the upgrade process to be pretty smooth, including for the UPMC portlets.
The upgrade time for libraries has greatly improved and continues to improve. :racehorse: 
Since you are passionate about keeping libraries up-to-date and secure, I highly recommend becoming more involved in the upgrade process. Pull requests are always welcome! :trophy: 
If you have additional questions or concerns, I would highly recommend discussing them with the uPortal steering committee. :speech_balloon: 

---

### Long Version

> it's a huge work that "no-one" is doing regularly

Library updates have been coming more regularly.
Shout-out to @JoshBrudnak, @kdjohnson, @CR96, and @cousquer  for their efforts in this area, thanks! :bowing_man: 
If you would like to see a further increased pace of upgrades, opening pull requests is a great way to assist in the upgrade effort.
[selection_003](https://github.com/Jasig/resource-server/compare)


> our team works shouldn't work with this last commit.

They should. :white_check_mark: 

All major release series of jquery still exist in the resource server.
The only thing that needs to update is the minor/patch versions of libraries used.

The upgrade from backbone 0.9 to 1.0 should be non-breaking [[reference](http://backbonejs.org/#changelog)]

> create two temporary branches in order to facilitate the dependencies updates of uPortal ecosystem and the updates of old projects by the community

One of the nice parts about uPortal is the modularity.
Portals levaraging the removed libraries can continue to use the current version of resource server, until they are ready to upgrade.

**Comment by auxepaul** [comment]
@ChristianMurphy 
I work since 6 years with @cousquer at the UPMC on our portal project with uPortal ecosystem (v4.x). We try to use the same version of js libs in the portal, portlets,... and we share these works with the community. I'm the last one to have updated jquery, jquery-ui... on uPortal 4.1.x, 4.2.x and 4.3.x and I have pushed lots of improvement in resource-server.
So I understand why you do that but I think it's premature to remove all old dependencies before to start the updates of projects that depend on resource-server project :
With this commit, It will be difficult to share this for example : https://github.com/Jasig/resource-server/pull/63#issuecomment-319073580 because resource-server release won't contain together : jquery 1.11.3, jqueryui 1.10.4 with boostrap 3.3.7. Have you try to update jqueryui before removing 1.10.x versions?

**Comment by drewwills** [comment]
@auxepaul, @ChristianMurphy, et al.,

Based on what I know (which is always imperfect, admittedly), removing libraries from the `resource-server-content` module should not be as big of an issue as it probably appears.

For portlets (or other projects) that use the resource server, there are some Best Practices we employ that give the portlet some resiliency to this sort of change.  Take a look at this example from the Calendar portlet:

  - https://github.com/Jasig/CalendarPortlet/blob/master/pom.xml#L843

The `resource-server-content` war is declared as a dependency of the portlet.  In Maven, that triggers the Maven Overlay behavior:  the contents of the resource-server war will be included in the Calendar portlet war.  The configs in the link are there to limit how much of the resource-server will be included it the Calendar.  It's really big, so you only want the things you're actually using.

If this practice is followed correctly, each portlet that uses an outdated (removed) library will actually have a copy of that library in its own webapp.  At runtime, the portlet will determine if the library is present in the resource-server.  If it is, great:  it will bring the library into the portlet markup from there;  if not, it will pull it from it's own copy.

**Comment by ChristianMurphy** [comment]
> we share these works with the community. [...] lots of improvement in resource-server.

And those efforts are appreciated! :bowing_man: Keep them coming! :+1: 

>  I think it's premature to remove all old dependencies before to start the updates of projects that depend on resource-server project :
> With this commit, It will be difficult to share this for example : #63 (comment) because resource-server release won't contain together : jquery 1.11.3, jqueryui 1.10.4 with boostrap 3.3.7.

See @drewwills comment above, overlays and resource-server automatically resolve this. :white_check_mark:
